### PR TITLE
Test for change to cache templates by site, not contents

### PR DESCRIPTION
**Pull request by littledan**
These tests are against a specification change based on discussion in
https://github.com/tc39/ecma262/issues/840

The tests here passed on SpiderMonkey but failed on other
implementations, which implement the current specification.

EDIT: They passed on a version of SpiderMonkey from February, and would likely fail on ToT.

### CODE REVIEW

File: test/language/expressions/tagged-template/cache-same-site.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+var previousObject = null;
+
+function runTemplate() {
+  tag`head${a}tail`;
+}
+
+runTemplate();
+firstObject = previousObject;
+
+assert(firstObject !== null);
```
Comment (leobalter) [inline test/language/expressions/tagged-template/cache-same-site.js]: assert(firstObject); should be fine, but you can actually get the proper value to test here, right?

```js
assert.sameValue(firstObject.length, 2);
assert.sameValue(firstObject[0], 'head');
assert.sameValue(firstObject[1], 'tail');
```

File: test/language/expressions/tagged-template/cache-differing-expressions-eval.js [hunk 2]
```diff
@@ -4,10 +4,8 @@
 es6id: 12.2.8
```
Comment (leobalter) [inline test/language/expressions/tagged-template/cache-differing-expressions-eval.js]: this behavior is not valid anymore for es6, these tags should be replaced by respective `esid` tags as in `esid: sec-gettemplateobject`

File: test/language/expressions/tagged-template/cache-same-site-top-level.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+let a = 1;
+for (let i = 0; i < 2; i++) {
+  tag`head${a}tail`;
+}
+
+assert.sameValue(templates.length, 2);
+
+assert.sameValue(
+  templates[0],
+  templates[1],
```
Comment (caitp) [inline test/language/expressions/tagged-template/cache-same-site-top-level.js]: I mentioned this on the normative issue, but I'm wondering how this is useful at all, for any reason.

I don't think we actually get any benefit at all from caching template objects, and this change would make it even less useful. I think we should just get rid of the caching entirely.

Comment (leobalter) [inline test/language/expressions/tagged-template/cache-same-site-top-level.js]: It's good to have the perspective from you, @caitp. It feels weird or wrong that we can observe we cache these references.

Comment (littledan) [inline test/language/expressions/tagged-template/cache-same-site-top-level.js]: JavaScript is a language that's based on objects with identity. If the system does any kind of caching of JS objects, it'll be observable exactly when it happens, for better or worse.

Comment (leobalter) [inline test/language/expressions/tagged-template/cache-same-site-top-level.js]: I don't see this being any problem for me as a web developer. It looks like a curious fact we can find the same reference.

I'm neutral to this change, if it helps runtimes, I'm +1 for it.

Comment (littledan) [inline test/language/expressions/tagged-template/cache-same-site-top-level.js]: For motivation, see https://github.com/tc39/ecma262/pull/890#issuecomment-293777826 and https://github.com/tc39/ecma262/pull/890#issuecomment-293924695

**Comment by littledan** [comment]
I will need to add a test for multiple parse sites that are in distinct functions, e.g., multiple closures with the same literal syntax, to test that they are cached the same. (In V8 terms, this means the cache lives on the SharedFunctionInfo, not the JSFunction.)

File: test/language/expressions/tagged-template/cache-same-site.js (hunk 1 above)
Comment (littledan) [inline test/language/expressions/tagged-template/cache-same-site.js]: I was going to say, "this is covered by other tests". But actually I can't find any other tests that cover it. Either way, I think validating what the template actually is is a subject for another, focused test. The assertion here is just to validate that the tag() function ran.

File: test/language/expressions/tagged-template/cache-differing-expressions-eval.js (hunk 2 above)
Comment (littledan) [inline test/language/expressions/tagged-template/cache-differing-expressions-eval.js]: Done

**Comment by leobalter** [comment]
@littledan this is here for a while. While I should give it a full review and resolve the merging conflicts, would you mind to review it as well?

**Comment by leobalter** [comment]
seems good to land for now. We can work on a follow up if necessary.

**Comment by littledan** [comment]
Sorry, I should've been more clear--the spec patch hasn't landed yet. I hope it should land soon!

**Comment by leobalter** [comment]
I think it's fine to have the tests considering we reached consensus in the respective PR? Similar to Stage 3 proposals, we have at least a reference. Otherwise I can revert it if necessary.

**Comment by littledan** [comment]
@leobalter Good analogy, that seems like a reasonable policy to me.

**Pull request merged**